pkg/systemd/overview-cards/cryptoPolicies.jsx

```diff
@@ -84,20 +85,38 @@ export const CryptoPolicyRow = () => {
     );
 };
 
-const setPolicy = (policy, setError, setInProgress) => {
+const setPolicy = async (policy, setError, setInProgress) => {
     setInProgress(true);
 
-    let promise;
-    if (policy === "FIPS") {
-        promise = cockpit.spawn(["fips-mode-setup", "--enable"], { superuser: "require", err: "message" });
-    } else {
-        promise = cockpit.spawn(["fips-mode-setup", "--disable"], { superuser: "require", err: "message" }).then(() =>
-            cockpit.spawn(["update-crypto-policies", "--set", policy], { superuser: "require", err: "message" }));
+    // If `fips-mode-setup` is available, we must use that (Fedora, RHEL 9); RHEL 10 dropped it, and adding fips=1 to
+    // the kernel command line automatically causes the other setup
+    async function setup_fips(fms_arg, kernelopt_args) {
+        try {
+            await cockpit.spawn(["fips-mode-setup", fms_arg], { superuser: "require", err: "message" });
+        } catch (error) {
+            if (error.problem === 'not-found') {
+                console.debug("fips-mode-setup not available, falling back to fips=1 kernel option");
+                await cockpit.script(kernelopt_sh, kernelopt_args, { superuser: "require", err: "message" });
+            } else {
+                throw error;
+            }
+        }
+    }
+
+    try {
+        if (policy === "FIPS") {
+            await setup_fips("--enable", ["set", "fips=1"]);
```

Comment:
There's a bit of a catch here. Switching to FIPS by adding `fips=1` to the kernel command line is only sufficient if `/boot` is not on a separate partition. If it is, adding `fips=1` without adding `boot=UUID=$boot_uuid` flag will cause the boot to fail because the initramfs can no longer figure out where to find the HMAC and kernel image to do the required integrity check.

So if you're implementing functionality to enable FIPS mode, I'd recommend you also add `boot=UUID=$(blkid --output value --match-tag UUID $(findmnt --first --noheadings -o SOURCE /boot))`.

Reply:
Meh -- that's exactly the kind of quirk that makes it hard to do and to document, i.e. what the `fips-mode-setup` helper was good at. By asking users to do that (manually, in Ansible deployments, bootc container builds, and what not), that'd reintroduce brittleness. If there is no choice here, why can't the dracut magic do that by itself?

Reply:
Done, it never offers enabling/disabling FIPS without `fips-mode-setup` now.

Reply:
@martinpitt Just to follow up on this, dracut can't set the kernel command line, so that part can't be automated.

You're right that we no longer support enabling FIPS mode after installation. It has all sorts of side effects to do that. For example, LUKS disk crypto uses PBKDF2 as key derivation function in FIPS mode during setup, but uses Argon2 otherwise. Systems that are installed outside of FIPS mode and switched afterwards then end up with either a non-FIPS key derivation function, or in the worst case, end up unbootable.

This is just one of the cases where behavior differs during installation depending on whether FIPS mode is enabled or not, so the choice to not offer enabling FIPS mode after installation is probably a good idea.

Reply:
Right, so I suppose stopping to support it is (or was, this landed already) the right move. Thanks for your review!